crates/schema/src/type_for_generate.rs

```diff
@@ -0,0 +1,516 @@
+
+    #[error("type {ty} contains identifier error {err}")]
+    NotValidIdentifier {
+        ty: PrettyAlgebraicType,
+        err: IdentifierError,
+    },
+}
+
+type Result<T> = std::result::Result<T, ErrorStream<ClientCodegenError>>;
+
+/// A typespace for generating client code.
+///
+/// The key difference is that this typespace stores only `AlgebraicTypeDef`s, not `AlgebraicType`s.
+/// We use the same `AlgebraicTypeRef`s from the original typespace.
+/// The difference is that `AlgebraicTypeRef`s ONLY point to `AlgebraicTypeDef`s.
+/// Chains of `AlgebraicTypeRef`s in the original `Typespace` are contracted to point to their ending `AlgebraicTypeDef`.
+#[derive(Debug, Clone)]
+pub struct TypespaceForGenerate {
+    defs: HashMap<AlgebraicTypeRef, AlgebraicTypeDef>,
+}
+
+impl TypespaceForGenerate {
+    /// Get the definitions of the typespace.
+    pub fn defs(&self) -> &HashMap<AlgebraicTypeRef, AlgebraicTypeDef> {
+        &self.defs
+    }
+
+    /// Build a `TypespaceForGenerate`.
+    ///
+    /// We're required to declare known definitions up front.
+    /// This is required for distinguishing between a use of the unit type, and a reference to a type declaration of a product type with no elements.
+    pub fn builder(
+        typespace: &Typespace,
+        known_definitions: impl IntoIterator<Item = AlgebraicTypeRef>,
+    ) -> TypespaceForGenerateBuilder<'_> {
+        TypespaceForGenerateBuilder {
+            typespace,
+            result: TypespaceForGenerate { defs: HashMap::new() },
+            known_definitions: known_definitions.into_iter().collect(),
+            uses: HashSet::new(),
+            known_uses: HashMap::new(),
+            currently_touching: HashSet::new(),
+        }
+    }
+}
+
+/// An algebraic type definition.
+#[derive(Debug, Clone)]
+pub enum AlgebraicTypeDef {
+    /// A product type declaration.
+    Product(ProductTypeDef),
+    /// A sum type declaration.
+    Sum(SumTypeDef),
+}
+
+/// A product type definition.
+#[derive(Debug, Clone)]
+pub struct ProductTypeDef {
+    /// The elements of the product type, in order.
+    pub elements: Box<[(Identifier, AlgebraicTypeUse)]>,
+}
+
+/// A sum type definition.
+#[derive(Debug, Clone)]
+pub struct SumTypeDef {
+    /// The variants of the sum type, in order.
+    pub variants: Box<[(Identifier, AlgebraicTypeUse)]>,
+}
+
+/// A use of an algebraic type.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum AlgebraicTypeUse {
+    /// A type where the definition is given by the typing context (`Typespace`).
+    /// In other words, this is defined by a pointer to another `AlgebraicType`.
+    /// An AlgebraicTypeUse must point to an `AlgebraicTypeDef`.
+    Ref(AlgebraicTypeRef),
+
+    /// The type of array values where elements are of a base type `elem_ty`.
+    /// Values [`AlgebraicValue::Array(array)`](crate::AlgebraicValue::Array) will have this type.
+    /// Stores an `Arc` for compression, the `Arc` is not semantically meaningful.
```

Comment:
I'm confused by this somewhat, why `Arc` and not plain `Box` if it's just for compression? Do we store reference-counted copies of the same type in several places?

Reply:
If multiple types contain identical sub-`AlgebraicTypeUse`s, they will use the same `Arc`, yeah. It's acyclic unless you pass through a Ref, just hash-consing. I can replace with Box if we don't wanna pay the Arc cost.

Reply:
Nah, makes sense. Although I guess `Rc` would work here too, since we never touch those structures concurrently?

Reply:
I was worried about some obscure thing demanding `Sync` later and got cold feet

Reply:
Fair enough, doesn't matter much anyway.